sno/gpkg_adapter.py

```diff
@@ -0,0 +1,257 @@
+import re
+from .dataset2 import Schema, ColumnSchema
+from osgeo.osr import SpatialReference
+
+
+GPKG_META_ITEMS = (
+    "gpkg_contents",
+    "gpkg_geometry_columns",
+    "gpkg_spatial_ref_sys",
+    "sqlite_table_info",
+    "gkpg_metadata",
+    "gpkg_metadata_reference",
+)
+
+
+def get_meta_item(dataset, path):
+    if path not in GPKG_META_ITEMS:
+        raise KeyError(f"Not a gpkg meta_item: {path}")
+
+    if path == "gpkg_contents":
+        return generate_gpkg_contents(dataset)
+    elif path == "gpkg_geometry_columns":
+        return generate_gpkg_geometry_columns(dataset)
+    elif path == "gpkg_spatial_ref_sys":
+        return generate_gpkg_spatial_ref_sys(dataset)
+    elif path == "sqlite_table_info":
+        return generate_sqlite_table_info(dataset)
+    elif path == "gpkg_metadata" or path == "gpkg_metadata_reference":
+        return None
+
+
+def generate_gpkg_contents(dataset2):
+    """Generate a gpkg_contents meta item from a dataset."""
+    gpkg_spatial_ref_sys = dataset2.get_meta_item("gpkg_spatial_ref_sys")
+    is_spatial = bool(gpkg_spatial_ref_sys)
+
+    result = {
+        "identifier": dataset2.get_meta_item("title"),
+        "description": dataset2.get_meta_item("description"),
+        "table_name": dataset2.tree.name,
+        "data_type": "features" if is_spatial else "attributes",
+    }
+    if is_spatial:
+        result["srs_id"] = gpkg_spatial_ref_sys[0]["srs_id"]
+    return result
+
+
+def generate_gpkg_geometry_columns(dataset2):
+    """Generate a gpkg_geometry_columns meta item from a dataset."""
+    geom_columns = _get_geometry_columns(dataset2.schema)
+    if not geom_columns:
+        return None
+
+    geometry_type = geom_columns[0].extra_type_info["geometryType"]
+    type_name, *zm = geometry_type.split(" ", 1)
+    zm = zm[0] if zm else ""
+    z = 1 if "Z" in zm else 0
+    m = 1 if "M" in zm else 0
+    gpkg_spatial_ref_sys = dataset2.get_meta_item("gpkg_spatial_ref_sys")
+    return {
+        "table_name": dataset2.tree.name,
+        "column_name": geom_columns[0].name,
+        "geometry_type_name": type_name,
+        "srs_id": gpkg_spatial_ref_sys[0]["srs_id"],
+        "z": z,
+        "m": m,
+    }
+
+
+def generate_gpkg_spatial_ref_sys(dataset):
+    """Generate a gpkg_spatial_ref_sys meta item from a dataset."""
+    geom_columns = _get_geometry_columns(dataset.schema)
+    if not geom_columns:
+        return []
+
+    srs_pathname = geom_columns[0].extra_type_info["geometrySRS"]
+    definition = dataset.get_srs_definition(srs_pathname)
+    return wkt_to_gpkg_spatial_ref_sys(definition)
+
+
+DEFAULT_GPKG_SPATIAL_REF_SYS = [
+    {
+        "srs_name": "Unknown CRS",
+        "definition": "",
+        "organization": "EPSG",
```

Comment:
`EPSG:0` isn't a thing. We should follow the approach of http://www.geopackage.org/spec/#r11 here, and use `NONE` if we really need a string.

Reply:
This preserves the previous import behaviour where we always assumed "EPSG", but srs_id would return 0 if we couldn't find a SpatialReference. But, happy to change it to "NONE"

Reply:
GPKG specifically has set values (NONE:0 and NONE:-1) that must be in every GPKG, but I think internally we should probably just not have a crs (& wkt) set? And the working copies can decide how to represent that (eg. PostGIS uses 0)